src/internet_identity/src/state/temp_keys.rs

```diff
@@ -0,0 +1,150 @@
+    /// cannot be used to authenticate (similarly to how the browser session key pair cannot be used to
+    /// authenticate if the actual delegated WebAuthn device is removed).
+    ///
+    /// In addition, the temporary key is also linked to an anchor so that even if the same device is
+    /// added to multiple anchors, it can only be used in the context of the anchor it was created for.
+    ///
+    /// Since temp keys can only be added during registration, the the max number of temp keys is
+    /// bounded by the registration rate limit.
+    temp_keys: HashMap<DeviceKey, TempKey>,
+
+    /// Heap to efficiently prune expired temp keys
+    expirations: BinaryHeap<TempKeyExpiration>,
+}
+
+impl TempKeys {
+    pub fn add_temp_key(
+        &mut self,
+        device_key: &DeviceKey,
+        anchor: AnchorNumber,
+        temp_key: Principal,
+    ) {
+        let tmp_key = TempKey {
+            principal: temp_key,
+            anchor,
+            expiration: time() + TEMP_KEY_EXPIRATION_NS,
+        };
+
+        self.expirations.push(TempKeyExpiration {
+            device_key: device_key.clone(),
+            expiration: tmp_key.expiration,
+        });
+        self.temp_keys.insert(device_key.clone(), tmp_key);
+    }
+
+    /// Removes the temporary key for the given device if it exists and is linked to the provided anchor.
+    pub fn remove_temp_key(&mut self, anchor: AnchorNumber, device_key: &DeviceKey) {
+        // we can skip the removal from expirations because there it will be removed
+        // during amortized clean-up operations
+        if let Some(temp_key) = self.temp_keys.get(device_key) {
+            if temp_key.anchor == anchor {
+                // Only remove temp key if the anchor matches
+                self.temp_keys.remove(device_key);
+            }
+        }
+    }
+
+    pub fn prune_expired_keys(&mut self) {
+        const MAX_TO_PRUNE: usize = 100;
+
+        let now = time();
+        for _ in 0..MAX_TO_PRUNE {
+            let Some(expiration) = self.expirations.peek() else {
+                break;
+            };
+
+            if expiration.expiration > now {
+                break;
+            }
+            self.temp_keys.remove(&expiration.device_key);
+            self.expirations.pop();
+        }
+    }
+
+    pub fn check_temp_key(
+        &self,
+        caller: &Principal,
+        device_key: &DeviceKey,
+        anchor: AnchorNumber,
+    ) -> Result<(), TempKeyError> {
+        let Some(temp_key) = self.temp_keys.get(device_key) else {
+            return Err(TempKeyError::NotFound);
+        };
+        if temp_key.expiration < time() {
+            return Err(TempKeyError::Expired);
+        }
+        if temp_key.anchor != anchor {
+            return Err(TempKeyError::AnchorMismatch);
+        }
+        if &temp_key.principal != caller {
+            return Err(TempKeyError::Invalid);
+        }
```

Comment:
Are these various error types read/used somewhere?

Can you always guarantee that an expired temp key will return `TempKeyError::Expired`? Won't it most likely return `NotFound`?

Reply:
Yeah, I thought the details would be more useful. But if needed, they can be added back later.